Support additional methods of account management with account plugins

chris-j-h opened the pull request:
This change adds support for account plugins to quorum and clef.  Account plugins can be used to provide alternative methods of managing accounts to the standard `keystore` file-based accounts.

Changes include:
* adds new `accounts.Pluggable` wallet and backend
* adds `geth account plugin` CLI commands
* enables usage of all existing account/signing-related APIs with plugin accounts (excluding `personal_newAccount` and `personal_importRawKey`)
* adds `plugin@account_newAccount` and `plugin@account_importRawKey` APIs
* extends `js` console with `plugin_account.newAccount` and `plugin_account.importRawKey`
* centralises account/key management docs (renames `Safeguarding Private Keys` section to `Account/Key Management`)

chris-j-h committed:
Support additional methods of account management with account plugins
This change adds support for account plugins to quorum and clef.  Account plugins can be used to provide alternative methods of managing accounts to the standard keystore file-based accounts.

Changes include:
* adds new accounts.Pluggable wallet and backend
* adds geth account plugin CLI commands
* enables usage of all existing account/signing-related APIs with plugin accounts (excluding personal_newAccount and personal_importRawKey)
* adds plugin@account_newAccount and plugin@account_importRawKey APIs
* extends js console with plugin_account.newAccount and plugin_account.importRawKey
* centralises account/key management docs (renames Safeguarding Private Keys section to Account/Key Management)

chris-j-h committed:
Merge branch 'master' into account-plugins

chris-j-h committed:
docs: Add plugin account API examples for clef

vsmk98 commented on internal/ethapi/api.go:
```diff
@@ -344,23 +345,55 @@ func (s *PrivateAccountAPI) UnlockAccount(ctx context.Context, addr common.Addre
+	case *pluggable.Backend:
+		return b.TimedUnlock(acct, password, duration)
+	case *keystore.KeyStore:
+		return b.TimedUnlock(acct, password, duration)
```
please add a `case` for `default` and return an error message. Otherwise for external signers like clef, this function will return `true` though the unlock operation would not have happened

chris-j-h replied:
Resolved by https://github.com/jpmorganchase/quorum/pull/1033/commits/e9d59669f6f34afcd2f3a50e6f3997ec2d2d7925

vsmk98 commented on internal/ethapi/api.go:
```diff
@@ -344,23 +345,55 @@ func (s *PrivateAccountAPI) UnlockAccount(ctx context.Context, addr common.Addre
+	case *pluggable.Backend:
+		return b.Lock(acct)
+	case *keystore.KeyStore:
+		return b.Lock(addr)
```
same as `unlockAccount`. Handle the default condition for any other type of signer and return error

chris-j-h replied:
Resolved by https://github.com/jpmorganchase/quorum/pull/1033/commits/e9d59669f6f34afcd2f3a50e6f3997ec2d2d7925

chris-j-h committed:
Error during lock/unlock if account is not keystore or plugin

chris-j-h committed:
Fix an intermittent seg viol when starting clef with account plugins
After oking the initial startup prompt with account plugins (and without the --nousb flag) there is intermittently a panic.
This occurs when the goroutine to start the USBListener (for usb wallets) (which attempts to open ALL wallets) is scheduled before the plugin wallet has been completely setup.
Completing the plugin wallet setup at the earliest possible stage (and before the USBListener goroutine is created) should resolve this error.

trung reviewed (dismissed):
LGTM

vsmk98 reviewed (dismissed):
LGTM

chris-j-h committed:
Merge branch 'master' into account-plugins

# Conflicts:
#	cmd/utils/flags.go
#	internal/web3ext/web3ext.go

chris-j-h committed:
Add account plugin API js console docs
